[feature] 카드 연관관계 매핑 및 조회기능 개발

sxi8022 opened the pull request:
## 관련 Issue

#52

## 변경 사항

- [x] 댓글 전체 조회 메소드 구현 (pageable 구현)
![image](https://github.com/proLmpa/NBC_2tWasSummer/assets/42510512/c657a966-e22c-4d7f-a132-7a1dbcfeb102)

-> 카드id에 따라 댓글 목록을 조회할 수 있도록 처리 
- [x] 라벨 검색(filter에서)

![필터검색_라벨id로 카드검색 조회](https://github.com/proLmpa/NBC_2tWasSummer/assets/42510512/48adf08e-46b8-41c1-baf4-464fc0990e64)

-> 라벨 테이블, 카드라벨 테이블, 카드 테이블을 조인하여 라벨id조건으로 검색 

 - [x] 카드목록 전체 조회 /api/cardLists

![카드전체조회](https://github.com/proLmpa/NBC_2tWasSummer/assets/42510512/0c6d4afc-9905-4330-bbd2-e368cb0166e6)

1. 보드의 키값을 requestParam으로 읽는다
2. 전체 카드정보를 조회하되 덱<카드목록>, 덱<카드목록> 이렇게 표시되도록 dto를 생성
 카드 상세 보기 할때, cardlabel 목록과, 댓글 목록도 표시되도록 개선

- [x] 덱과 연관관계 매핑 후 postman 으로 테스트


![덱생성후카드생성](https://github.com/proLmpa/NBC_2tWasSummer/assets/42510512/8c8f4292-d537-48ab-9de9-9ba9f41bd1e4)

-> 매핑 후 필요한 코드 작성 (ex : dto 객체)

## Todo List

테스트 코드 별도로 작성

## Check List

- [x] 포스트맨으로 체크해 보았나요?

## 트러블 슈팅

-  queryDsl 결과를 list<엔티티> 형태로 그대로 표시하다가 에러가 났었다. 
   -> 별도의 dto객체로 변환 
- postman 테스트 할때 request Dto 객체를 formData 에 넣어서 전송하다가 
   multipart 파일이 허용되지 않습니다. 라는 에러가 떠서 자료들을 찾아봤다. 시행착오를 겪으면서 테스트를 원할히 진행할 수 있었다. 
- label 과 cardlabel 테이블 매핑이 한쪽에서만 되어있어서 데이터를 repository에서 불러오는데 어려움을 겪었는데, 
 에러메시지가 명확하지 않아 해결하는데 시간이 걸렸다.

sxi8022 committed:
[refactor] Card 관련 연관관계 매핑 작업
- 댓글 전체 조회 기능 추가
- 댓글 crud 를 postman으로 테스트가 용이하게 변경

sxi8022 committed:
[refactor] Card 관련 연관관계 매핑 작업
- 덱과 연관관계 추가

sxi8022 committed:
[refactor] Card 관련 연관관계 매핑 작업
- 카드 상세 보기 할때, cardlabel 목록과, 댓글 목록도 표시되도록 개선

sxi8022 committed:
[refactor] Card 관련 연관관계 매핑 작업
- 보드값에 따라 카드 전체조회가 되도록 기능 구현

sxi8022 committed:
[refactor] Card 관련 연관관계 매핑 작업
- 라벨값에 따라 카드 전체조회가 되도록 기능 구현(필터검색)
- WebSecurityConfig 접근 허용이 되는 주소 추가
- 라벨 엔티티에 관계 매핑
- 소스코드 정리

JisooPyo commented:
![image](https://github.com/proLmpa/NBC_2tWasSummer/assets/130378232/b836cc8e-4d5a-4064-aafa-7a37c63de88a)

여기가 다 card라고 되어 있어서 어플리케이션을 실행하니까 이 부분에서 오류가 나는 것 같습니다..! 아직 진행 중인 부분이신거죠?

proLmpa commented on src/main/java/com/example/itwassummer/card/repository/CardRepositoryImpl.java:
```diff
@@ -28,4 +38,75 @@ public int changeOrder(Card selectCard, long order) {
         .execute();
     return 1;
```
이 메서드를 사용하는 Service 메서드를 확인해보니 굳이 값을 반환해주지 않아도 되는 거 같아서 return 타입을 void로 바꿔도 될 것 같습니다.

sxi8022 replied:
네 수정했습니다.!

proLmpa commented on src/main/java/com/example/itwassummer/card/controller/CardController.java:
```diff
@@ -116,17 +134,36 @@ public ResponseEntity moveCard(
     return new ResponseEntity<>(result, HttpStatus.OK);
   }
 
```
카드를 기존 덱에서 다른 덱으로 옮기는 기능은 아직 구현 안된 건가요?

sxi8022 replied:
네 그 부분은 구현하지 않았습니다. api를 새로 정해서 만들어야 할거 같아요.

proLmpa commented on src/main/java/com/example/itwassummer/comment/service/CommentService.java:
```diff
@@ -2,18 +2,29 @@
 
-    String createComment(Long cardId, CommentCreateRequestDto requestDto, User user);
+    /**
+     * 카드 상세 조회
```
여기 메서드 설명 제목들이 전부 똑같아용

sxi8022 replied:
수정 완료했습니다.

proLmpa commented on src/main/java/com/example/itwassummer/card/controller/CardController.java:
```diff
@@ -116,17 +134,36 @@ public ResponseEntity moveCard(
   @GetMapping("/cards/{cardId}/comments")
   @ResponseBody
-  public ResponseEntity list(
+  public ResponseEntity commentList(
```
labelId에 따른 댓글 조회 기능은 CardController 보다는 CommentController로 옮기는 게 좋을 것 같습니다.
Mapping 요청도 '/comments/{cardId}'로 바꿔서요!

sxi8022 replied:
![image](https://github.com/proLmpa/NBC_2tWasSummer/assets/42510512/90500ef7-5e74-4a3b-bae6-bde9ab110a62)

cardId에 따라 댓글을 조회하고 있습니다. 
피드백 대로 
comment controller로 옮기고 
mapping 요청도 바꾸었습니다.

proLmpa commented on src/main/java/com/example/itwassummer/comment/dto/CommentCreateRequestDto.java:
```diff
@@ -7,4 +7,5 @@
 @AllArgsConstructor
 public class CommentCreateRequestDto {
     private String content;
+    private Long cardId;
```
cardId는 이미 Controller에서 @PathVariable을 통해 받았기 때문에 굳이 추가할 필요는 없습니다.

sxi8022 replied:
![image](https://github.com/proLmpa/NBC_2tWasSummer/assets/42510512/fd942a79-a662-4994-8017-6af98a64be4f)

cardId는 현재 requestDto에 담아서 사용중입니다. api명세서에서는 cardId를 requestParam으로 담는 거라고 이해했습니다.

proLmpa replied:
저희가 작성한 API에서 cardId는 PathVariable로 받고, deckId만 RequestParam으로 받는 걸로 적혀 있습니단

proLmpa commented on src/main/java/com/example/itwassummer/comment/controller/CommentController.java:
```diff
@@ -23,14 +23,13 @@ public class CommentController {
+        String result = commentService.createComment(requestDto, userDetails.getUser());
 
         return ResponseEntity.ok().body(new ApiResponseDto(HttpStatus.OK.value(), result));
         //issue ResponseEntity 반환을 어떻게 하는 게 베스트일지?
```
어.. 포스트맨에 결과 출력만 된다면 편하신 대로??

sxi8022 replied:
해당 주석은 삭제하였습니다.

proLmpa commented on src/main/java/com/example/itwassummer/cardlabel/entity/CardLabel.java:
```diff
@@ -26,7 +30,6 @@ public CardLabel(Card card, Label label) {
         this.label = label;
```
이거 29번째 줄에 card의 cardLabels 에 추가하는 절차를 넣으셨는데 이게 양방향 매핑되어 있어서 추가하는 코드 없이도 알아서 들어가도록 동작할 거예요

29번째 줄 주석 처리해서 실행해보시고 잘 동작한다면 삭제해주세요

sxi8022 replied:
![image](https://github.com/proLmpa/NBC_2tWasSummer/assets/42510512/bd57f518-d6a6-4709-a5f7-11f7dac5ef70)

등록 잘 되는지 확인하고 해당 소스코드는 지웠습니다.

proLmpa commented on src/main/java/com/example/itwassummer/card/dto/CardListResponseDto.java:
```diff
@@ -0,0 +1,25 @@
+import lombok.Getter;
+
+@Getter
+public class CardListResponseDto {
```
지금 ResponseDto가 메서드 별로 다른데 보면 중복으로 출력되는 필드들이 있어서 그냥 ResponseDto를 하나로 합치는 게 좋을 것 같습니다.

CardViewResponseDto가 Card의 필드들을 모두 포함하고 있어서 해당 dto를 CardResponseDto로 바꿔 통일 시키면 좋겠습니다.

sxi8022 replied:
CardViewResponseDto 에 첨부파일 하고 날짜 정보가 들어가서 
저는 원하는 정보만 출력한다는 의미로 여러개 만들었습니다. 
프론트에 원하는 정보만 내려준다는 의미에서는 괜찮지 않을까요? 카드id를 받아오면 상세화면은 다 뿌려준다
는 생각을 했습니다.

proLmpa replied:
필요한 정보 가공을 백에서 할지, 프론트에서 할지에 대한 차이인 것 같네요. 그럼 이건 그대로 두죠

proLmpa reviewed (commented):
Card가 연관된 class가 못해도 4~5개 되다 보니까 많이 복잡하고 이해하는 데 시간이 좀 걸렸습니다.

피드백입니다.

sxi8022 commented:
> ![image](https://user-images.githubusercontent.com/130378232/260024319-b836cc8e-4d5a-4064-aafa-7a37c63de88a.png)
> 
> 여기가 다 card라고 되어 있어서 어플리케이션을 실행하니까 이 부분에서 오류가 나는 것 같습니다..! 아직 진행 중인 부분이신거죠?
제가 로그인할때나 포스트맨으로 테스트할때 에러가 나진 않는데 
어느부분에서 오류가 발생하나요?

sxi8022 reviewed (commented):
리뷰에 대해 조치했습니다.

sxi8022 committed:
[refactor] Card 관련 연관관계 매핑 작업
- 댓글 목록 조회 기능 CommentController 로 이동 및 매핑 경로 수정
- 카드 정렬 순서 변경 repository 메소드의 리턴형을 void로 변경
- 주석 수정, 삭제
- cardlabel 엔티티 수정

sxi8022 committed:
[refactor] Card 관련 연관관계 매핑 작업
- 카드를 다른 덱으로 이동

sxi8022 committed:
Merge branch 'dev' into feature/52/cardMapping

sxi8022 committed:
[refactor] Git Action 테스트 반영으로 인해 카드 관련 테스트가 동작하지 않아 코드 주석 처리

proLmpa reviewed (approved):
주말에도 코드 보시느라 고생하셨어요!~

저 오늘 발표 전까지 Controller 테스트 돌리는 법 찾아보고 얻는 게 있으면 바로 공유 드릴게요 :D
